### PR TITLE
Add support for Gitea

### PR DESCRIPTION
**Pull request by MagicRB**
This pull request is just there to publicize the code I'm working on, before being ready it'll get probably more cleaned up and merged with the Gitea backend pull request.

### CODE REVIEW

File: buildbot_nix/__init__.py [hunk 1]
```diff
@@ -239,7 +262,9 @@ def __init__(self, **kwargs: Any) -> None:
     @defer.inlineCallbacks
     def run(self) -> Generator[Any, object, Any]:
         # run `nix build`
-        cmd: remotecommand.RemoteCommand = yield self.makeRemoteShellCommand()
```
Comment (Mic92) [inline buildbot_nix/__init__.py:242]: What was the issue with the previous line?

File: buildbot_nix/__init__.py [hunk 2]
```diff
@@ -224,7 +245,9 @@ def run(self) -> Generator[Any, object, Any]:
         error = self.getProperty("error")
         attr = self.getProperty("attr")
         # show eval error
-        error_log: Log = yield self.addLog("nix_error")
+        # TODO why doesn't type information pass through yield again?
+        error_log_: object  = yield self.addLog("nix_error")
```
Comment (Mic92) [inline buildbot_nix/__init__.py:249]: same question as above.

File: buildbot_nix/__init__.py [hunk 3]
```diff
@@ -145,17 +164,22 @@ class NixEvalCommand(buildstep.ShellMixin, steps.BuildStep):
...
         super().__init__(**kwargs)
+        self.project = project
         self.observer = logobserver.BufferLogObserver()
         self.addLogObserver("stdio", self.observer)
         self.supported_systems = supported_systems
 
     @defer.inlineCallbacks
     def run(self) -> Generator[Any, object, Any]:
         # run nix-eval-jobs --flake .#checks to generate the dict of stages
-        cmd: remotecommand.RemoteCommand = yield self.makeRemoteShellCommand()
```
Comment (Mic92) [inline buildbot_nix/__init__.py:158]: Same question as above.

File: buildbot_nix/__init__.py (hunk 1 above)
Comment (MagicRB) [reply buildbot_nix/__init__.py:242]: All of your questions, types, typechecking seemed to be broken for me, soemhow through a yield it wouldnt type check, my guess is that a yield can return an error value. I wanted to remove all the type errors so its easier for me to write (and for the future). I can revert these bits no problem if so desired

File: buildbot_nix/__init__.py [hunk 4]
```diff
@@ -347,8 +322,9 @@ def run_vc(
             return res
 
         # The code below is a modified version of Git.run_vc
-        self.stdio_log: Log = yield self.addLogForRemoteCommands("stdio")
-
+        # TODO why doesn't type information pass through yield again?
+        stdio_log_: object = yield self.addLogForRemoteCommands("stdio")
```
Comment (Mic92) [inline buildbot_nix/__init__.py:326]: Same question as above.

File: buildbot_nix/__init__.py (hunk 1 above)
Comment (Mic92) [reply buildbot_nix/__init__.py:242]: What typechecker are you using?

Comment (MagicRB) [reply buildbot_nix/__init__.py:242]: pyright

Comment (MagicRB) [reply buildbot_nix/__init__.py:242]: I will try mypy then and clean up these workarounds 👍. Sorry for the mess

Comment (Mic92) [reply buildbot_nix/__init__.py:242]: Alright. You can use `nix fmt` to get a local type check.

**Comment by MagicRB** [comment]
Ready for review! I verified all functionality to be working against a fork of this repository at [https://github.com/magicrb/buildbot-nix](https://github.com/magicrb/buildbot-nix) and [https://codeberg.org/magic_rb/buildbot-nix](https://codeberg.org/magic_rb/buildbot-nix). No regressions have been introduced as far as I can tell. The following section includes notes that I took while working on this.

# Initial Setup

-   confused about how the GH integration works. The docs aren't great.
-   <https://github.com/settings/tokens/new> took me ages to find
-   the gh app must be installed and oauth enabled
-   git is required
-   force eval manually is broken, properties do not get populated
-   will only consider PRs or default branch, no way to make it consider other branches

# Removing GitHub Assumption

-   gonna end up using <https://github.com/lab132/buildbot-gitea>, thank god I don't have to implement this myself;
-   <https://github.com/microsoft/pyright/issues/6000> so therefore `buildbot.plugins.util.BuilderConfig` is invalid?
    -   `from buildbot.config.builder import BuilderConfig` works

# Adding Gitea

-   Gitea's API is **very** similar to GitHub's, especially the responses. Only thing different so far is that Gitea doesn't include `topics` in the repo listing, but that is easy to splice in.
-   It is very unclear whether Gitea includes repos of orgs a user belongs to in the `/api/v1/user/repos` endpoint, if not it's easy to splice in again
-   what happens if the same project `owner/repo` is on Gitea and GitHub
-   the reporters are stupid, they will trigger all for every repo, i dont know how to make them not trigger for the incorrect one, maybe subclassing and a filter?s

# Cleanup and Finalizing

-   still don't know how to resolve this

    ```python
    # TODO Gitea doesn't include this information
    return False # self.data["owner"]["type"] == "Organization"
    ```

    Gitea really doesn't seem to expose that information, [the API docs](https://forgejo.dev/api/swagger#/repository/repoGet) agree.


Lastly:
mypy is a much weaker type checker in my experience, it would miss many cases of errors that pyright would report, but I did undo the weird type casting since we're using mypy.

**Comment by mannp** [comment]
Very happy for this, thank you. Out of interest, would this support private forgejo/gitea repos or public only?

**Comment by MagicRB** [comment]
> Very happy for this, thank you. Out of interest, would this support private forgejo/gitea repos or public only?

If the buildbot instance can access the instance, https works, then it should be fine I think. It is untested though.

**Comment by MagicRB** [comment]
oh yeah, there are a few unresolved things: 
1. reporters will trigger for both gitea and github, both of them. The gitea one will skip the report gracefully, the github one will throw an exception. But it seems to be harmless
2. The Gitea plugin doesnt support avatars
3. Gitea doesn't seem to distinguish between organizations and users in its API, so `belongs_to_org` is a noop and always returns `False`
4. buildbot doesn't support multiple authentication plugins at once, the default is `github`

File: buildbot_nix/github_projects.py [hunk 5]
```diff
@@ -1,93 +1,229 @@
...
+            )
+            return util.FAILURE
+
+
+@dataclass
+class GithubConfig:
+    oauth_id: str
+
+    # TODO unused
+    buildbot_user: str
```
Comment (zowoq) [inline buildbot_nix/github_projects.py:71]: Does `TODO unused` mean `buildbot_user` can be removed?

Comment (MagicRB) [reply buildbot_nix/github_projects.py:71]: At the time I wasn't a 100% sure, but now I am, and yes it can

Comment (zowoq) [reply buildbot_nix/github_projects.py:71]: Thanks, I'll send a PR.

**Pull request merged**